src/esp/gfx/MaterialData.h

```diff
@@ -15,21 +15,79 @@
 namespace esp {
 namespace gfx {
 
-struct MaterialData {};
+enum class MaterialDataType {
+  /** @brief default, does not represent any material */
+  None = 0,
+  /** @brief phong material data */
+  Phong = 1,
+  /** @brief Physically based rendering (PBR) material data */
+  Pbr = 2,
+};
+
+struct MaterialData {
+  MaterialData(MaterialDataType _type = MaterialDataType::None) : type(_type){};
+
+  MaterialDataType type{MaterialDataType::None};
+
+  bool perVertexObjectId = false;
+
+  // construct it using the default constructor. NO initial values, such as
+  // identity matrix
+  Magnum::Matrix3 textureMatrix;
+};
 
 struct PhongMaterialData : public MaterialData {
+  PhongMaterialData() : MaterialData(MaterialDataType::Phong){};
+
   Magnum::Float shininess = 80.f;
   Magnum::Color4 ambientColor{0.1};
   Magnum::Color4 diffuseColor{0.7};
   Magnum::Color4 specularColor{0.2};
-  Magnum::Matrix3 textureMatrix;
   Magnum::GL::Texture2D *ambientTexture = nullptr, *diffuseTexture = nullptr,
                         *specularTexture = nullptr, *normalTexture = nullptr;
-  bool perVertexObjectId = false, vertexColored = false;
+  bool vertexColored = false;
 
   ESP_SMART_POINTERS(PhongMaterialData)
 };
 
+struct PbrMaterialData : public MaterialData {
+  PbrMaterialData() : MaterialData(MaterialDataType::Pbr){};
+
+  // material with default magic values:
+  // when both the material property and the texture (e.g., roughness and
+  // roughness texture) are NOT presented, use these default magic values;
+  Magnum::Color4 baseColor{0.7f};
+  Magnum::Float roughness = 0.9f;
+  Magnum::Float metallic = 0.1f;
+  Magnum::Color3 emissiveColor{0.0f};
+  Magnum::Float normalTextureScale = 1.0f;
+
+  Magnum::GL::Texture2D* baseColorTexture = nullptr;
+  Magnum::GL::Texture2D* normalTexture = nullptr;
+  Magnum::GL::Texture2D* emissiveTexture = nullptr;
+
+  // Question:
+  // What if e.g., both metallicTexture and nonRoughnessMEtallicTexture exist?
+  // Answer:
+  // Such conflict will be handled in the PbrShader based on the priority of the
+  // textures. See generateCorrectFlags() in PbrShader for details.
+
+  // an independent, NOT packed texture to store "metallic" (in R
+  // channel)
+  Magnum::GL::Texture2D* metallicTexture = nullptr;
+  // an independent, NOT packed texture to store "roughness" (in R channel)
+  Magnum::GL::Texture2D* roughnessTexture = nullptr;
+  // a packed texture to store "roughness" (in G channel) and "metallic" (in B
+  // channel)
+  Magnum::GL::Texture2D* noneRoughnessMetallicTexture = nullptr;
+  // a packed texture to store "occlusion"(in R channel), "roughness" (in G
+  // channel) and "metallic" (in B channel)
+  Magnum::GL::Texture2D* occlusionRoughnessMetallicTexture = nullptr;
+  // TODO:
+  // AO texture
+  ESP_SMART_POINTERS(PbrMaterialData)
+};
```

Comment:
Yeah so because this class will inevitably grow further as support for new map packing and material properties gets added, this is where I originally thought Magnum's [Trade::MaterialData](https://doc.magnum.graphics/magnum/classMagnum_1_1Trade_1_1MaterialData.html) could get used, instead of maintaining a downstream copy of it. It supports everything listed here, the colors and properties can be specified via builtin attributes and the texture pointers [via custom attributes](https://doc.magnum.graphics/magnum/classMagnum_1_1Trade_1_1MaterialData.html#Trade-MaterialData-populating-custom).

Opinions? @aclegg3 @eundersander